orcas/l1l2cassandra.go

```diff
@@ -0,0 +1,411 @@
+// Copyright 2018 Criteo
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package orcas
+
+import (
+	"errors"
+	"log"
+
+	"github.com/netflix/rend/common"
+	"github.com/netflix/rend/handlers"
+	"github.com/netflix/rend/metrics"
+	"github.com/netflix/rend/orcas"
+	"github.com/netflix/rend/protocol"
+	"github.com/netflix/rend/timer"
+)
+
+var (
+	ErrL1L2SetFailed   = errors.New("ERROR Both L1 & L2 Set failed")
+	ErrL2OnlySetFailed = errors.New("ERROR L2Only Set failed")
+)
+
+type L1L2CassandraOrca struct {
+	l1  handlers.Handler
+	l2  handlers.Handler
+	res protocol.Responder
+}
+
+func L1L2Cassandra(l1, l2 handlers.Handler, res protocol.Responder) orcas.Orca {
+	return &L1L2CassandraOrca{
+		l1:  l1,
+		l2:  l2,
+		res: res,
+	}
+}
+
+// This func aims at providing an async method to fill L2 after L1 set is ack'ed
+func (l *L1L2CassandraOrca) SetL2Only(req common.SetRequest) error {
+	// Set to L2
+	metrics.IncCounter(orcas.MetricCmdSetL2)
+	start := timer.Now()
+	errL2 := l.l2.Set(req)
+
+	metrics.ObserveHist(orcas.HistSetL2, timer.Since(start))
+
+	// Return an error if data is not stored in L2
+	if errL2 != nil {
+		metrics.IncCounter(orcas.MetricCmdSetErrorsL2)
+		return ErrL2OnlySetFailed
+	}
+
+	metrics.IncCounter(orcas.MetricCmdSetSuccessL2)
+	// return code is ignored by goroutine caller (Set func)
+	return errL2
+}
+
+func (l *L1L2CassandraOrca) Set(req common.SetRequest) error {
+	//log.Println("set", string(req.Key))
+
+	// Set to L1 first
+	metrics.IncCounter(orcas.MetricCmdSetL1)
+	start := timer.Now()
+	errL1 := l.l1.Set(req)
+
+	metrics.ObserveHist(orcas.HistSetL1, timer.Since(start))
+
+	if errL1 != nil {
+		metrics.IncCounter(orcas.MetricCmdSetErrorsL1)
+	} else {
+		// Successful write to L1 is a completed write !
+		metrics.IncCounter(orcas.MetricCmdSetSuccessL1)
+		metrics.IncCounter(orcas.MetricCmdSetSuccess)
+		// Set L2 asynchronously. As L1 is OK, we don't check L2 success.
+		go l.SetL2Only(req)
```

Comment:
I am not sure that spawning a go routine will scale (don't know go much)
but ok for a v0